[5.9][move-only] Fix SIL representation and SILOptimizer to preserve value deinits.

atrick opened the pull request:
--- CCC ---

Explanation:

Many basic SIL passes were silently deleting the struct and enum deinits:

- SROA
- Mem2Reg
- RLE
- DSE
- FunctionSignatureOpts
- LowerAggregates
- SILCombine
- EarlyCodeMotion
- ReleaseHoisting
- many passes that rely on ARC analysis (RCIndentity)

Major Changes:

- add SIL verification for drop_deinit
- fix ownership lowering to handle drop_deinit
- disable deinit devirtualization (until the optimizer is tested)
- fix 'shouldExpand' to recognize value deinits; impacts many passes
- fix dead function elimination to collect and preserve value deinits
- fix RCIdentity to recognize value deinits; impacts many passes
- fix SILCombine special cases to preserve value deinits
- fix TypeLowering to preserve value deinits; mainly fixes LowerAggregates

Scope of Issue: In some cases, deinitialization would run recursively, hanging the system. In other cases, a struct's deinit would silently fail to run when the struct was a member within another type. In the most trivial cases, with no type composition, and earlier MoveOnlyDeinitInsertion pass was hiding the problem.

Risk: Low risk to regular copyable code. And non-copyable was already broken.

PR to main: [move-only] Fix SIL representation and SILOptimizer to preserve value deinits #66314
https://github.com/apple/swift/pull/66314

Reviewed By: Erik Eckstein and Joe Groff

Resolves:

rdar://109849028 ([move-only] LowerAggregateInstrs eliminates struct deinitialization)

rdar://109863801 ([move-only] DCE removes drop_deinit, so user-defined deinitializers call themselves recursively at -O)

rdar://110079028 ([move-only] Finish SIL support for drop_deinit)

atrick committed:
[move-only] Fix SILOptimizer expansion to preserve deinits.

Many basic SIL passes were silently deleting the struct deinit:
- SROA
- Mem2Reg
- RLE
- DSE
- FunctionSignatureOpts
- LowerAggregates

Fixes rdar://109849028 ([move-only] LowerAggregateInstrs eliminates struct deinitialization)

(cherry picked from commit 4a37b00cae0c3802959a566795644ccc74eb4b1e)

atrick committed:
Fix drop_deinit side effects.

rdar://109863801 ([move-only] DCE removes drop_deinit, so user-defined
deinitializers call themselves recursively at -O)

(cherry picked from commit af49f62d6364ce76077ea23d93914f42f8d547d2)

atrick committed:
[move-only] Fix drop_deinit OSSA lowering

drop_deinit ultimately only affects the semantics of its
destroy_value. Avoid generating releases for destroys in which the
deinit has been dropped. Instead, individually release the members.

(cherry picked from commit a6f65f07c588e1f815bd161df3252e786191ddcd)

atrick committed:
[move-only] Comment emitMoveOnlyMemberDestruction drop_deinit

(cherry picked from commit cc0c44c4ac941e038b60c43d336e2b53a2cded65)

atrick committed:
[move-only] Verify drop_deinit

drop_deinit only exists in ownership SIL. Remove IRGen support.

A drop_deinit can only ever be destroyed or destructured.

A destructure of a struct-with-deinit requires a drop_deinit operand.

(cherry picked from commit 3efd122f04302a51bf77656d5bc264c785aa2be4)

atrick committed:
[move-only] Rename MoveOnlyDeinitInsertion

to MoveOnlyDeinitDevirtualization

(cherry picked from commit 9855edd9c96a427c7c8b407fa1ca4a57db64feb5)

atrick committed:
[move-only] Disable move-only devirtualization.

It is not needed for correctness and hides most of the deinit related
optimizer bugs.

(cherry picked from commit 84bee24a77e0bfb85ed83318efc14bce6a58e1ff)

atrick committed:
[move-only] Fix dead code elimination to preserve value deinits.

(cherry picked from commit 1dcc240e15db2a48eab314b68bb9df0557c30c52)

atrick committed:
[move-only] Fix SILOptimizer code motion to preserve value deinits

Multiple code motion and ARC related passes were removing struct/enum
deinits.

Passes fixed include:
- SILCombine
- EarlyCodeMotion
- ReleaseHoisting
- *many* passes that rely on ARC analysis (RCIndentity)

(cherry picked from commit 6756ed7dfb0969b4a2eb2565022c1bfba0249cda)

atrick committed:
[move-only] Fix SILCombine to preserve value deinit

Fix a special case in visitReleaseValueInst for enum-with-deinit.

(cherry picked from commit 58f96a93b5ae539d185fa50fcc73a66c5649a81c)

atrick committed:
[move-only] Fix TypeLowering to preserve value deinits

MoveOnlyLoadableStruct should only lower to memberwise destroys if it
has no deinit.

(cherry picked from commit 986cd42c82dedc285fce11e1f3525d577516a6cd)

atrick committed:
[move-only] Update tests for drop_deinit lowering.

(cherry picked from commit 14f591788f461eba48a0d3fab6eb0c09c904e857)

atrick commented:
@swift-ci test

eeckstein reviewed (commented):
optimizer changes LGTM

eeckstein commented on lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp:
```diff
@@ -1082,7 +1086,7 @@ SILInstruction *SILCombiner::visitReleaseValueInst(ReleaseValueInst *RVI) {
     // retain_value of an enum_inst where we know that it has a payload can be
     // reduced to a retain_value on the payload.
-    if (EI->hasOperand()) {
+    if (EI->hasOperand() && !hasValueDeinit(EI)) {
```
This `hasValueDeinit` is redundant because it's already checked in line 1078

atrick replied:
That redundant check is a result of multiple fixes that were merged into this PR. I'll cleanup the extra check on main first to avoid merge conflicts.

atrick replied:
That cleanup is now in https://github.com/apple/swift/pull/66314

eeckstein commented on lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp:
```diff
@@ -1074,6 +1074,10 @@ SILInstruction *SILCombiner::visitReleaseValueInst(ReleaseValueInst *RVI) {
   SILType OperandTy = Operand->getType();
 
+  // Do not remove a release that calls a value deinit.
+  if (hasValueDeinit(OperandTy))
```
This check also needs to be done in `isTrivial` above. If the operand type is a struct without a deinit, but one of the struct fields has a deinit, the release_value will be removed. E.g.
```
  struct WithDeinit : ~Copyable {
    deinit {}
  }
  struct Outer {
    var wd: WithDeinit
  }

  ...
  release_value %1 : $Outer
```

atrick replied:
Right! That issue was fixed here, but you didn't see it because I rebased this PR on top of Kavon's
```
commit e20243219042458ff256bb6f88e53cc086cb3736 (public/5.9-fix-silcombine)
Author: Andrew Trick <atrick@apple.com>
Date:   Thu May 25 16:14:16 2023

    Fix SILCombine to preserve a release of a move-only type

    The release needs to be preserved in case a user-defined deinit is
    present in the released type. Checking for move-only is slightly
    conservative.

    Fixes rdar://109846094 ([move-only] SILCombine eliminates struct deinitialization)

    (cherry picked from commit dc974e8c00d923474905d9c41d6b71bc045751d9)
```

eeckstein reviewed (changes requested):
I think I found a potential problem inSILCombine's `isTrivial` function

atrick commented:
@swift-ci test

atrick committed:
Disable SILGen/discard.swift on Windows.

The -Onone passes fail to remove struct_extract, so the CHECK
lines will fail.

(cherry picked from commit d55ca8a78d65c056ca673120d14f06630cfc9dfb)

kavon commented on lib/SILGen/SILGenDestructor.cpp:
```diff
@@ -497,113 +497,59 @@ void SILGenFunction::emitClassMemberDestruction(ManagedValue selfValue,
 
 void SILGenFunction::emitMoveOnlyMemberDestruction(SILValue selfValue,
```
Between `ownershipmodeleliminatorvisitor::splitDestroy` and `silgenfunction::emitMoveOnlyMemberDestruction` I'm not really sure who's suppose to be expanding the destruction now.

Seems like the intention is that we emit this in SILGen:
```
%dd = drop_deinit self
destroy_value %dd
```

and then when we lower out of OSSA we expand that `destroy(drop_deinit)` like we use to in SILGen with memberwise-destruction, like:

```
%elms = destructure_struct self
destroy %elm1
...
```

This model seems nicer; it leaves the decision of how to destroy the value up to SIL, and uses just the marker `drop_deinit` to know that a release is not how to do it since it triggers deinit.

atrick replied:
@kavon Right. Unfortunately `drop_deinit` can also be used on address types, in which case it means something significantly different. The interesting features of `drop_deinit` are the OSSA/lifetime semantics. The address form doesn't have any of those semantics. It's just a placeholder to indicate that we need to be careful when promoting the address to OSSA. We should to add a separate `drop_deinit_addr`, but I didn't get to that in this PR.

A couple things to note:

SILGen *could* legally expand the `destroy_value` into a `destructure_struct`, but that would be silly and non-canonical.

The optimizer can promote an address `drop_deinit` to a value (mem2reg), but when doing so, we need to check the same conditions present in the SILVerifier... it can only be used to destroy or destructure.

eeckstein reviewed (approved):
lgtm!

atrick commented:
@swift-ci clean test macOS

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test
